### PR TITLE
gh-112066: Add `PyDict_SetDefaultRef` function.

### PR DESCRIPTION
**Pull request by colesbury**
The `PyDict_SetDefaultRef` function is similar to `PyDict_SetDefault`, but returns a strong reference through the optional `**result` pointer instead of a borrowed reference.

* Issue: gh-112066

----
:books: Documentation preview :books:: https://cpython-previews--112123.org.readthedocs.build/

### CODE REVIEW

**Comment by colesbury** [comment]
cc @zooba, @encukou, and @serhiy-storchaka in case you are interested in this

**Comment by zooba** [comment]
My only comment is the same as for the recently added pop functions: we should not use these functions as precedent for their API design (specifically how the return value is handled). When we come to decide which approach we want to standardise for the C API in general, we aren't going to be forced into choosing this design just because it was used during the period between deciding-to-decide and actually deciding.

But since there's a chance that the other recently added API could also become "warts", best to align with those since they're all on the same type. There's nothing gained by deliberate inconsistency when the functions show up right next to each other.

**Review by serhiy-storchaka — COMMENTED** [review]
Very nice. The code looks good.

But I worry that all exit branches simply add `Py_NewRef()` for `PyDict_SetDefaultRef()`. It means that `PyDict_SetDefaultRef()` is not principally different from `PyDict_SetDefault()` combined with `Py_XNewRef()`. The question arises about the need to add a new function.

Could you please create a separate PR that replaces all `PyDict_SetDefault()` with `PyDict_SetDefaultRef()` to show what benefit it could bring.

File: Doc/c-api/dict.rst [hunk 1]
```diff
@@ -174,6 +174,20 @@ Dictionary Objects
...
+
+   Inserts *default_value* into the dictionary *p* with a key of *key* if the
+   key is not already present in the dictionary. If *result* is not ``NULL``,
+   then *\*result* is set to a :term:`strong reference` to either
+   *default_value*, if the key was not present, or the existing value, if *key*
+   was already present in the dictionary.
+   Returns ``1`` if the key was present and *default_value* was not inserted,
+   ``0`` if the key was not present and *default_value* was inserted, or
+   ``-1`` on failure.
+
```
Comment (encukou) [inline Doc/c-api/dict.rst]:
```suggestion
   Returns ``1`` if the key was present and *default_value* was not inserted,
   or ``0`` if the key was not present and *default_value* was inserted.
   On failure, returns ``-1``, sets an exception, and sets ``*result``
   to ``NULL``.

   For clarity: if you have a strong reference to *default_value* before
   calling this function, then after it returns, you hold a strong reference
   to both *default_value* and *\*result* (if it's not ``NULL``).
   These may refer to the same object: in that case you hold two separate
   references to it.
```

**Comment by encukou** [comment]
> we should not use these functions as precedent for their API design

IMO, if the C-API WG is formed and decides soon enough, this API and all its uses in CPython might be adjusted. That might affect in-flight PRs.

---

> `PyDict_SetDefaultRef()` is not principally different from `PyDict_SetDefault()` combined with `Py_XNewRef()`

With nogil, the latter is incorrect. The reference can become invalid (and the object deallocated) between the `PyDict_SetDefault()` and `Py_XNewRef()` calls.
(That doesn't apply if you know no other threads can mutate the dict, but future changes can easily invalidate assumptions like that.)

**Comment by colesbury** [comment]
As @encukou wrote, returning a strong reference is important in nogil builds and can't be safely simulated with a subsequent `Py_INCREF()`.

I'll put up a subsequent PR that replaces internal `PyDict_SetDefault` calls with `PyDict_SetDefaultRef`.

**Comment by colesbury** [comment]
Are there any outstanding concerns with this change @encukou @serhiy-storchaka?

**Review by serhiy-storchaka — COMMENTED** [review]
Thank you for #112211 (yet one nice number!).

Please add a What's New entry. The rest LGTM.

**Comment by colesbury** [comment]
I've added the "What's New" entry

**Review by serhiy-storchaka — APPROVED** [review]
LGTM.

**Comment by colesbury** [comment]
@serhiy-storchaka - would you please merge this?

**Comment by colesbury** [comment]
Looks like the C API WG approved the API. @serhiy-storchaka, can we merge this PR now?

**Comment by serhiy-storchaka** [comment]
One more issue needs to be resolved. Wait another week. :wink:

**Comment by erlend-aasland** [comment]
> One more issue needs to be resolved. Wait another week. 😉

Which issue?

**Comment by serhiy-storchaka** [comment]
The issue I was talking about was the ongoing vote to promote @colesbury as a core developer. I think it would be nice if Sam merge his PR himself. Welcome on the board!

**Pull request merged**